<a href="https://colab.research.google.com/github/ArunK-ML/ML-Model---Diabetes_Prediction---DataSet/blob/main/Diabetes_Prediction_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [25]:
import pandas as pd

In [26]:
df = pd.read_csv("/content/diabetes_prediction_dataset.csv")

df

gender   age  hypertension  heart_disease smoking_history    bmi  \
0      Female  80.0             0              1           never  25.19   
1      Female  54.0             0              0         No Info  27.32   
2        Male  28.0             0              0           never  27.32   
3      Female  36.0             0              0         current  23.45   
4        Male  76.0             1              1         current  20.14   
...       ...   ...           ...            ...             ...    ...   
99995  Female  80.0             0              0         No Info  27.32   
99996  Female   2.0             0              0         No Info  17.37   
99997    Male  66.0             0              0          former  27.83   
99998  Female  24.0             0              0           never  35.42   
99999  Female  57.0             0              0         current  22.43   

       HbA1c_level  blood_glucose_level  diabetes  
0              6.6                  140         0  
1              6.6                   80         0  
2              5.7                  158         0  
3              5.0                  155         0  
4              4.8                  155         0  
...            ...                  ...       ...  
99995          6.2                   90         0  
99996          6.5                  100         0  
99997          5.7                  155         0  
99998          4.0                  100         0  
99999          6.6                   90         0  

[100000 rows x 9 columns]

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 9 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   gender               100000 non-null  object 
 1   age                  100000 non-null  float64
 2   hypertension         100000 non-null  int64  
 3   heart_disease        100000 non-null  int64  
 4   smoking_history      100000 non-null  object 
 5   bmi                  100000 non-null  float64
 6   HbA1c_level          100000 non-null  float64
 7   blood_glucose_level  100000 non-null  int64  
 8   diabetes             100000 non-null  int64  
dtypes: float64(3), int64(4), object(2)
memory usage: 6.9+ MB


In [28]:
# Count Duplicate entry

df.duplicated().sum()

np.int64(3854)

In [29]:
# dropping duplicates

df.drop_duplicates(keep = "last",inplace = True)

df

gender   age  hypertension  heart_disease smoking_history    bmi  \
0      Female  80.0             0              1           never  25.19   
2        Male  28.0             0              0           never  27.32   
3      Female  36.0             0              0         current  23.45   
4        Male  76.0             1              1         current  20.14   
5      Female  20.0             0              0           never  27.32   
...       ...   ...           ...            ...             ...    ...   
99995  Female  80.0             0              0         No Info  27.32   
99996  Female   2.0             0              0         No Info  17.37   
99997    Male  66.0             0              0          former  27.83   
99998  Female  24.0             0              0           never  35.42   
99999  Female  57.0             0              0         current  22.43   

       HbA1c_level  blood_glucose_level  diabetes  
0              6.6                  140         0  
2              5.7                  158         0  
3              5.0                  155         0  
4              4.8                  155         0  
5              6.6                   85         0  
...            ...                  ...       ...  
99995          6.2                   90         0  
99996          6.5                  100         0  
99997          5.7                  155         0  
99998          4.0                  100         0  
99999          6.6                   90         0  

[96146 rows x 9 columns]

In [30]:
df["gender"].unique()

array(['Female', 'Male', 'Other'], dtype=object)

In [31]:
df["smoking_history"].unique()

array(['never', 'current', 'No Info', 'former', 'ever', 'not current'],
      dtype=object)

In [32]:
# To change the object into numbers based on ASCII value

from sklearn.preprocessing import LabelEncoder

code = LabelEncoder()

for i in df.columns:
  if df[i].dtype=="object":
    df[i]= code.fit_transform(df[i]) # Encoding

df

gender   age  hypertension  heart_disease  smoking_history    bmi  \
0           0  80.0             0              1                4  25.19   
2           1  28.0             0              0                4  27.32   
3           0  36.0             0              0                1  23.45   
4           1  76.0             1              1                1  20.14   
5           0  20.0             0              0                4  27.32   
...       ...   ...           ...            ...              ...    ...   
99995       0  80.0             0              0                0  27.32   
99996       0   2.0             0              0                0  17.37   
99997       1  66.0             0              0                3  27.83   
99998       0  24.0             0              0                4  35.42   
99999       0  57.0             0              0                1  22.43   

       HbA1c_level  blood_glucose_level  diabetes  
0              6.6                  140         0  
2              5.7                  158         0  
3              5.0                  155         0  
4              4.8                  155         0  
5              6.6                   85         0  
...            ...                  ...       ...  
99995          6.2                   90         0  
99996          6.5                  100         0  
99997          5.7                  155         0  
99998          4.0                  100         0  
99999          6.6                   90         0  

[96146 rows x 9 columns]

**Encoding**

**Gender**

Female = 0,
Male = 1,
Other = 2

**smoking_history**

'never' = 4 , 'No Info' = 0 , 'current' = 1, 'former' = 3, 'ever' = 2, 'not current' = 5

In [33]:
# Where independent = x , dependent = y

x = df.iloc[:,:-1] # x = defining independent columns with all the row
y = df.iloc[:,-1]  # y = defining dependent columns with all the row

In [34]:
x.columns

Index(['gender', 'age', 'hypertension', 'heart_disease', 'smoking_history',
       'bmi', 'HbA1c_level', 'blood_glucose_level'],
      dtype='object')

In [35]:
y.name

'diabetes'

In [36]:
# Check the shape of x and y

x.shape, y.shape

((96146, 8), (96146,))

In [37]:
# Count the diabetes

df["diabetes"].value_counts()

diabetes
0    87664
1     8482
Name: count, dtype: int64

In [38]:
df.shape

(96146, 9)

# **Machine Learning**

Total number of rows : 96146

The total rows take 100% and we splite :

*   30% for test - 28844 approx. ----> x_test + y_test
*   70% for Train - 67302 approx. ----> x_train + y_train


# Is for understanding
Diabetes Count :

*   "0" -> 87664
*   "1" -> 8482

We split 70% and 30% in each diabetes count ("0" & "1")

For "0" -> 30% -> 2545  ---> x_test

For "0" -> 70% -> 5937  ---> x_train

For "1" -> 30% -> 26299  ---> y_test

For "1" -> 70% -> 61365  ---> y_train


Adding 30% -> 2545 + 26299 = 28844 &

70% -> 5937 + 61365 = 67302

In [39]:
from sklearn.model_selection import train_test_split

x_train , x_test , y_train , y_test = train_test_split(x,y,test_size = 0.3)

x_train.shape, x_test.shape, y_train.shape, y_test.shape

((67302, 8), (28844, 8), (67302,), (28844,))

In [40]:
y_train.value_counts()

diabetes
0    61331
1     5971
Name: count, dtype: int64

In [41]:
#Create Model or Train the machine

from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier().fit(x_train,y_train)

In [42]:
# Sample Predicting

import numpy as np
vb1 = np.array([[1,35,0,0,4,32.64,9.3,180]])

vb2 = np.array([[1,75,1,1,0,32.64,6.3,180]])


In [43]:
model.predict(vb1)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


array([1])

In [44]:
model.predict(vb2)

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


array([1])

In [45]:
y_pred = model.predict(x_test)
y_pred

array([0, 0, 0, ..., 0, 0, 1])

In [46]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report

confusion_matrix(y_test,y_pred)

array([[25570,   763],
       [  639,  1872]])

In [47]:
accuracy_score(y_test,y_pred)

0.9513937040632368

In [48]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.98      0.97      0.97     26333
           1       0.71      0.75      0.73      2511

    accuracy                           0.95     28844
   macro avg       0.84      0.86      0.85     28844
weighted avg       0.95      0.95      0.95     28844

